fuel/datasets/hdf5.py

```diff
@@ -266,98 +476,93 @@ def unsorted_fancy_index(request, indexable):
+        # file handle. However, since `load` is also called during the lifetime
+        # of a dataset (e.g. if load_in_memory = True), we don't want to
+        # accidentally overwrite the reference to a potential external file
+        # handle, hence this check.
+        if not hasattr(self, '_external_file_handle'):
+            self.external_file_handle = None
 
-    def load_axis_labels(self):
         self._out_of_memory_open()
         handle = self._file_handle
-        axis_labels = {}
-        for source_name in handle:
-            if source_name in self.vlen_sources:
-                axis_labels[source_name] = (
-                    (handle[source_name].dims[0].label,) +
-                    tuple(label.decode('utf8') for label in
-                          handle[source_name].dims[0]['shape_labels']))
-            else:
-                axis_labels[source_name] = tuple(
-                    dim.label for dim in handle[source_name].dims)
-        self._out_of_memory_close()
-        return axis_labels
-
-    @property
-    def available_splits(self):
-        return tuple(self.split_dict.keys())
 
-    @property
-    def provides_sources(self):
-        return tuple(self.split_dict[self.which_set].keys())
-
-    @property
-    def vlen_sources(self):
-        if not hasattr(self, '_vlen_sources'):
-            self._out_of_memory_open()
-            handle = self._file_handle
-            vlen_sources = []
-            for source_name in self.sources:
-                source = handle[source_name]
-                if len(source.dims) > 0 and 'shapes' in source.dims[0]:
-                    if len(source.dims) > 1:
-                        raise ValueError('Variable-length sources must have '
-                                         'only one dimension.')
-                    vlen_sources.append(source_name)
-            self._vlen_sources = tuple(vlen_sources)
-            self._out_of_memory_close()
-        return self._vlen_sources
-
-    @property
-    def subsets(self):
-        if not hasattr(self, '_subsets'):
-            subsets = [self._subset_template for source in self.sources]
-            num_examples = None
-            for i, source_name in enumerate(self.sources):
-                start, stop = self.split_dict[self.which_set][source_name][:2]
-                subset = subsets[i]
+        # Load subset slices / indices
+        subsets = []
+        if len(self.which_set) > 1:
+            indices = defaultdict(list)
+            for split in self.which_set:
+                split_start_stop = self.get_start_stop(handle, split)
+                split_indices = self.get_indices(handle, split)
+                for source_name in self.sources:
+                    if source_name in split_indices:
+                        ind = split_indices[source_name]
+                    else:
+                        ind = list(range(*split_start_stop[source_name]))
+                    indices[source_name] = sorted(
```

Comment:
The more I think about this issue the more it seems to me that your solution is a reasonable one. Just mention this procedure in the `which_set` docstring, and I guess it will be all right.